### DISCUSSION TITLE
AUC calculation with trapezoidal rule

### DISCUSSION THREAD

**Discussion by gabrielmcg44** [ClickHouse/ClickHouse#65192] [Ideas]
Hey folks! Firstly thanks for the awesome work with clickhouse, it is being really awesome to use it at work.

I was comparing the AUC I get from python's sklearn and the one I get with clickhouse arrayAUC and there is a small (but not negligible) difference. As I was studying the causes I understood that clickhouse approximates the Area Under Curve with rectangles, meanwhile sklearn approximate the areas by trapezoidal rule.

The result was that for AUCs above 0.5 clickhouse got greater values than sklearn while for AUCs under 0.5 it got smaller values.

I might be wrong but this seems to be the source of the divergence. Here follows an image to ilustrate what I am saying:

<img width="446" alt="Screenshot 2024-06-12 at 13 09 32" src="https://github.com/ClickHouse/ClickHouse/assets/52935333/7f307065-06a8-4bb6-bce3-52fd306e8df3">

It seems trapezoidal rule is a more standard way of calculating this, is there a reason to clickhouse do this with rectangles? Is it something related to performance?

It would be awesome to be able to reproduce the same results we have with sklearn in Clickhouse if there was an option to use trapezoidal rule :)

References:
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.auc.html
https://clickhouse.com/docs/en/sql-reference/functions/array-functions#arrayauc
https://github.com/ClickHouse/ClickHouse/blob/fbb36fc7380e7ed3ff41dbb758e1922694519a4c/src/Functions/array/arrayAUC.cpp#L118

**Comment by UnamedRus** [comment]
> is there a reason to clickhouse do this with rectangles?

No, person X contributed arraAUC, which he developed for it's own needs. And looks like it was rectangles mode.

**Reply by gabrielmcg44** [reply]
Understood! Could I open a PR for this feature when I have some time then? How does it work?

**Reply by UnamedRus** [reply]
> Could I open a PR for this feature

Yes

You can look on implementation of https://github.com/ClickHouse/ClickHouse/blob/fbb36fc7380e7ed3ff41dbb758e1922694519a4c/src/Functions/array/arrayFill.cpp#L125

How to make 2 functions with small difference in behavior

**Reply by UnamedRus** [reply]
Makes sense to have multiple functions

arrayAUCRiemann (or arrayAUCRectangle(Left? to be more precise, as there are 4 different options exist https://en.wikipedia.org/wiki/Riemann_sum) ), which is current implementation

arrayAUCTrapezoidal, is your implementation

arrayAUC will be just an alias to arrayAUCRectangleLeft

We have similar issue for quantileExact (quantileExactHigh, quantileExactLow, quantileExactExclusive, quantileExactInclusive)

How to make an alias 
https://github.com/ClickHouse/ClickHouse/blob/905495e59b7025674a0cba94fca57beeac5c6651/src/Functions/log.cpp#L38

**Reply by gabrielmcg44** [reply]
Thank you for the code pointers. It helps a lot. I will try to setup/implement things here to open a PR

**Comment by gabrielmcg44** [comment]
While I was trying to implement it, I could take a deeper look in the current implementation and noticed that it was not trying to do any approximation, instead it was trying to calculate the exact Area Under Curve. 

But when there were duplicate scores, it was creating multiple points for the same score. This is what was causing the divergences between clickhouse and sklearn AUCs, also, different table ordering would affect AUC. So I opened this PR with a fix:

https://github.com/ClickHouse/ClickHouse/pull/65840
